zerver/views/registration.py

```diff
@@ -181,6 +188,13 @@ def accounts_register(request: HttpRequest) -> HttpResponse:
         except ValidationError:
             return redirect_to_email_login_url(email)
 
+        if settings.BILLING_ENABLED:
+            try:
+                check_spare_licenses_available_for_adding_new_users(realm, 1)
+            except LicenseLimitError:
+                send_user_unable_to_signup_message_to_signup_notification_stream(realm, email)
```

Comment:
It's kinda hard to read that we call `send_user_unable_to_signup_message_to_signup_notification_stream` in several places; I'm worried we're e.g. missing the `zproject/backends.py` user creation code paths.

Perhaps as part of cleaning this up, `check_spare_licenses_available_for_adding_new_users` could handle calling that function when it finds a shortage, at least if the `for_new_account=True` parameter is passed.